CBL-4237 : Pass cert chain to the root_cert_locator()

pasin opened the pull request:
**Background for CBL-4234:**

In LiteCore PublicKey+Apple.mm’s findSigningRootCert() for TLS CA cert validation, a SecTrust is created with only a leaf cert without its intermediate certs. As a result, when calling SecTrustEvaluateWithError() to evaluate the SecTrust (a technique to get all certs including the root CA cert used for verifying a CA signed cert), there is a warning error logged as follows:

```
[TLS] error: SecTrustEvaluateWithError failed: Error Domain=NSOSStatusErrorDomain Code=-25318 "errKCCreateChainFailed / errSecCreateChainFailed:  / The attempt to create a certificate chain failed."
```

Code Reference : https://github.com/couchbase/couchbase-lite-core/blob/master/Crypto/PublicKey%2BApple.mm#L809-L813

This issue was found while testing CBL-C Replicator on Mac with a Capella endpoint. Even though there is an error logged but a root CA cert of the Capella endpoint was still found by the SecTrustEvaluateWithError() for some reason. To avoid unexpected issue when intermediate certs are required for validation, we should include all of the certs received from the remote server for validation.

**Fix:**

This commit is constructing and passing the cert chain (in PEM format) to the root_cert_locator(). The passed cert chain, will eventually be passed to the findSigningRootCert() that Apple platform can use when constructing an array of SecCertRef for creating a SecTrust. 

Note that there will be a corresponding fix in LiteCore to create a SecTrust with the cert chain as well.

pasin committed:
CBL-4237 : Pass cert chain to the root_cert_locator()

Background for CBL-4234:

In LiteCore PublicKey+Apple.mm’s findSigningRootCert() for TLS CA cert validation, a SecTrust is created with only a leaf cert without its intermediate certs. As a result, when calling SecTrustEvaluateWithError() to evaluate the SecTrust (a technique to get all certs including the root CA cert used for verifying a CA signed cert), there is an error logged as follows:

```
[TLS] error: SecTrustEvaluateWithError failed: Error Domain=NSOSStatusErrorDomain Code=-25318 "errKCCreateChainFailed / errSecCreateChainFailed:  / The attempt to create a certificate chain failed."
```

Fix:

This commit is constructing and passing the cert chain (in PEM format) to the root_cert_locator(). The passed cert chain, will eventually be passed to the findSigningRootCert() that Apple platform can use when constructing an array of SecCertRef for creating a SecTrust. Note that there will be a corresponding fix in LiteCore regarding this as well.

snej commented on src/mbedtls_context.cpp:
```diff
@@ -587,7 +588,30 @@ namespace sockpp {
+        string certData;
+        for (auto crt = child; crt; crt = crt->next) {
+            size_t olen = 0;
+            unsigned char buf[4096];
```
Is it OK to use a fixed size buffer? What if the cert is bigger?

pasin replied:
I have just changed the buffer size to 10K to match what LiteCore is using in mbedUtils.cc.

pasin replied:
I have searched the code that uses mbedUtils.convertToPEM() which uses a fixed size buffer of 10K bytes, but I haven't seen the code paths that actually convert cert (DER) into PEM especially for the cert validation purpose. In general the 10K buffer should be enough but as there is no size limit for the TLS certs, to be sure, I have changed the code to use dynamic size buffer and retry again when the initial buffer size is too small.

bmeike reviewed (approved):
Shame you have to copy it twice.

Looks about right, to me.

pasin committed:
Change max buffer size to 10000 to match LiteCore’s mbedUtils.cc

Per my chat with Jens, changing the max buffer size to 10000 seems to safe enough especially if we’re using that limit elsewhere.

Reference : https://github.com/couchbase/couchbase-lite-core/blob/8aff93a70d94593e7ec9c770a2648b123473065c/Crypto/mbedUtils.cc#L130-L143

jianminzhao commented:
Please make another PR to merge it to release/3.1

pasin committed:
Use dynamic buffer instead of fixed size buffer

Use vector<char> as a dynamic buffer with an initial buffer size of 10000 which should be big enough for most cases. When the buffer size is not big enough, resize the buffer with the min buffer size set in olen out param by the mbedtls_pem_write_buffer function and retry again.
